rs/src/gstd/events.rs

```diff
@@ -37,26 +39,162 @@ fn with_optimized_event_encode<T, E: SailsEvent, F: FnOnce(&[u8]) -> T>(
+        unimplemented!(
+            "`emit_event` is implemented only for the wasm32 architecture and the std future"
+        )
+    }
+}
+
+#[cfg(feature = "ethexe")]
+impl<T: super::EthEvent> EventEmitter<T> {
+    /// Emits an event for the Ethexe program.
+    #[cfg(target_arch = "wasm32")]
+    pub fn emit_eth_event(&mut self, event: T) -> crate::errors::Result<()> {
+        super::ethexe::__emit_eth_event(event)
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    #[cfg(not(feature = "std"))]
+    pub fn emit_eth_event(&mut self, _event: T) -> crate::errors::Result<()> {
+        unimplemented!(
+            "`emit_eth_event` is implemented only for the wasm32 architecture and the std future"
+        )
+    }
 }
 
-#[doc(hidden)]
-#[cfg(target_arch = "wasm32")]
-pub fn __emit_event_with_route<TEvents>(route: &[u8], event: TEvents) -> crate::errors::Result<()>
-where
-    TEvents: SailsEvent,
-{
-    with_optimized_event_encode(route, event, |payload| {
-        gstd::msg::send_bytes(gstd::ActorId::zero(), payload, 0)?;
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(feature = "std")]
+impl<T: 'static> EventEmitter<T> {
+    /// Emits an event.
+    pub fn emit_event(&mut self, event: T) -> crate::errors::Result<()> {
+        event_registry::push_event(self.route, event);
         Ok(())
-    })
+    }
+
+    #[cfg(feature = "ethexe")]
+    pub fn emit_eth_event(&mut self, event: T) -> crate::errors::Result<()> {
+        event_registry::push_event(self.route, event);
+        Ok(())
+    }
+
+    /// Takes the events emitted for this route and returns them as a `Vec<T>`.
+    pub fn take_events(&mut self) -> crate::Vec<T> {
+        event_registry::take_events(self.route).unwrap_or_else(|| crate::Vec::new())
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(feature = "std")]
+mod event_registry {
+    use core::any::{Any, TypeId};
+    use std::{boxed::Box, collections::BTreeMap, sync::Mutex, vec::Vec};
+
+    type Key = (&'static [u8], TypeId);
+
+    std::thread_local! {
+        /// thread-local registry mapping `(key, TypeId)` -> boxed `Vec<T>`
+        static ROUTE_EVENTS: Mutex<BTreeMap<Key, Box<dyn Any>>> = Mutex::new(BTreeMap::new());
+    }
+
+    /// Push a `value: T` onto the `Vec<T>` stored under `key`.
+    /// If none exists yet, we create a `Vec<T>` for that `(key,TypeId::of::<T>())`.
+    pub(super) fn push_event<T: 'static>(key: &'static [u8], value: T) {
+        ROUTE_EVENTS.with(|mtx| {
+            let mut map = mtx.lock().expect("failed to lock ROUTE_EVENTS mutex");
+            let slot = map
+                .entry((key, TypeId::of::<T>()))
+                .or_insert_with(|| Box::new(Vec::<T>::new()));
+            // SAFETY: we just inserted a Box<Vec<T>> for exactly this TypeId,
+            // so downcast_mut must succeed.
+            let vec: &mut Vec<T> = slot
+                .downcast_mut::<Vec<T>>()
+                .expect("type mismatch in route-events registry");
+            vec.push(value);
```

Comment:
Is it possible to get rid of downcasting?
```suggestion
                .or_insert_with(|| {
                    let mut vec = Vec::<T>::new();
                    vec.push(value);

                    Box::new(vec)
                });
```